<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Making and Utilizing TxDb Objects</title>
<style type="text/css">
/**
 * Prism.s theme ported from highlight.js's xcode style
 */
pre code {
  padding: 1em;
}
.token.comment {
  color: #007400;
}
.token.punctuation {
  color: #999;
}
.token.tag,
.token.selector {
  color: #aa0d91;
}
.token.boolean,
.token.number,
.token.constant,
.token.symbol {
  color: #1c00cf;
}
.token.property,
.token.attr-name,
.token.string,
.token.char,
.token.builtin {
  color: #c41a16;
}
.token.inserted {
  background-color: #ccffd8;
}
.token.deleted {
  background-color: #ffebe9;
}
.token.operator,
.token.entity,
.token.url,
.language-css .token.string,
.style .token.string {
  color: #9a6e3a;
}
.token.atrule,
.token.attr-value,
.token.keyword {
  color: #836c28;
}
.token.function,
.token.class-name {
  color: #DD4A68;
}
.token.regex,
.token.important,
.token.variable {
  color: #5c2699;
}
.token.important,
.token.bold {
  font-weight: bold;
}
.token.italic {
  font-style: italic;
}
</style>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  box-sizing: border-box;
}
body, .footnotes, code { font-size: .9em; }
li li { font-size: .95em; }
*, *:before, *:after {
  box-sizing: inherit;
}
pre, img { max-width: 100%; }
pre, pre:hover {
  white-space: pre-wrap;
  word-break: break-all;
}
pre code {
  display: block;
  overflow-x: auto;
}
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre) > code, code[class] { background-color: #F8F8F8; }
code.language-undefined, pre > code:not([class]) {
  background-color: inherit;
  border: 1px solid #eee;
}
table {
  margin: auto;
  border-top: 1px solid #666;
}
table thead th { border-bottom: 1px solid #ddd; }
th, td { padding: 5px; }
thead, tfoot, tr:nth-child(even) { background: #eee; }
blockquote {
  color: #666;
  margin: 0;
  padding-left: 1em;
  border-left: 0.5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC .numbered li { list-style: none; }
#TOC .numbered { padding-left: 0; }
#TOC .numbered ul { padding-left: 1em; }
table, .body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.footnote-ref a::before { content: "["; }
.footnote-ref a::after { content: "]"; }
section.footnotes::before {
  content: "";
  display: block;
  max-width: 20em;
}

@media print {
  body {
    font-size: 12pt;
    max-width: 100%;
  }
  tr, img { page-break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  pre { white-space: pre; }
}
</style>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex/dist/katex.min.css">
</head>
<body>
<div class="frontmatter">
<div class="title"><h1>Making and Utilizing TxDb Objects</h1></div>
<div class="author"><h2>Marc Carlson, Patrick Aboyoun, Hervé Pagès, Seth Falcon, Martin Morgan</h2></div>
<div class="date"><h3>Wednesday, March 13, 2024</h3></div>
</div>
<div class="body">
<h2 id="introduction">Introduction</h2>
<p>The <code>GenomicFeatures</code> package retrieves and manages
transcript-related features from the UCSC Genome
Bioinformatics\footnote{(<a href="https://genome.ucsc.edu/)%7D">https://genome.ucsc.edu/)}</a> and
BioMart\footnote{(<a href="http://www.biomart.org/)%7D">http://www.biomart.org/)}</a> data resources. The
package is useful for ChIP-chip, ChIP-seq, and RNA-seq analyses.</p>
<pre><code class="language-r">suppressPackageStartupMessages(library('GenomicFeatures'))
</code></pre>
<h2 id="txdb-objects"><code>TxDb</code> Objects</h2>
<p>The <code>GenomicFeatures</code> package uses <code>TxDb</code>
objects to store transcript metadata. This class maps the 5’ and 3’
untranslated regions (UTRs), protein coding sequences (CDSs) and exons
for a set of mRNA transcripts to their associated
genome. <code>TxDb</code> objects have numerous accessors functions to
allow such features to be retrieved individually or grouped together
in a way that reflects the underlying biology.</p>
<p>All <code>TxDb</code> objects are backed by a SQLite database that
manages genomic locations and the relationships between pre-processed
mRNA transcripts, exons, protein coding sequences, and their related
gene identifiers.</p>
<h2 id="retrieving-data-from-txdb-objects">Retrieving Data from <code>TxDb</code> objects</h2>
<h3 id="loading-transcript-data">Loading Transcript Data</h3>
<p>There are two ways that users can load pre-existing data to generate a
<code>TxDb</code> object.  One method is to use the
<code>loadDb</code> method to load the object directly from an
appropriate .sqlite database file.</p>
<p>Here we are loading a previously created <code>TxDb</code> object
based on UCSC known gene data.  This database only contains a small
subset of the possible annotations for human and is only included to
demonstrate and test the functionality of the
<code>GenomicFeatures</code> package as a demonstration.</p>
<pre><code class="language-r">samplefile &lt;- system.file(&quot;extdata&quot;, &quot;hg19_knownGene_sample.sqlite&quot;,
                          package=&quot;GenomicFeatures&quot;)
txdb &lt;- loadDb(samplefile)
txdb
</code></pre>
<pre><code>## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg19
## # Organism: Homo sapiens
## # UCSC Table: knownGene
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: no
## # miRBase build ID: NA
## # transcript_nrow: 178
## # exon_nrow: 620
## # cds_nrow: 523
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2014-10-08 10:31:15 -0700 (Wed, 08 Oct 2014)
## # GenomicFeatures version at creation time: 1.17.21
## # RSQLite version at creation time: 0.11.4
## # DBSCHEMAVERSION: 1.0
</code></pre>
<p>In this case, the <code>TxDb</code> object has been returned by
the <code>loadDb</code> method.</p>
<p>More commonly however, we expect that users will just load a
TxDb annotation package like this:</p>
<pre><code class="language-r">library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene #shorthand (for convenience)
txdb
</code></pre>
<pre><code>## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg19
## # Organism: Homo sapiens
## # Taxonomy ID: 9606
## # UCSC Table: knownGene
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: yes
## # miRBase build ID: GRCh37
## # transcript_nrow: 82960
## # exon_nrow: 289969
## # cds_nrow: 237533
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2015-10-07 18:11:28 +0000 (Wed, 07 Oct 2015)
## # GenomicFeatures version at creation time: 1.21.30
## # RSQLite version at creation time: 1.0.0
## # DBSCHEMAVERSION: 1.1
</code></pre>
<p>Loading the package like this will also create a <code>TxDb</code>
object, and by default that object will have the same name as the
package itself.</p>
<h3 id="pre-filtering-data-based-on-chromosomes">Pre-filtering data based on Chromosomes</h3>
<p>It is possible to filter the data that is returned from a
<code>TxDb</code> object based on it’s chromosome.  This can be a
useful way to limit the things that are returned if you are only
interested in studying a handful of chromosomes.</p>
<p>To determine which chromosomes are currently active, use the
<code>seqlevels</code> method.  For example:</p>
<pre><code class="language-r">head(seqlevels(txdb))
</code></pre>
<pre><code>## [1] &quot;chr1&quot; &quot;chr2&quot; &quot;chr3&quot; &quot;chr4&quot; &quot;chr5&quot; &quot;chr6&quot;
</code></pre>
<p>Will tell you all the chromosomes that are active for the
TxDb.Hsapiens.UCSC.hg19.knownGene <code>TxDb</code> object (by
default it will be all of them).</p>
<p>If you then wanted to only set Chromosome 1 to be active you could do
it like this:</p>
<pre><code class="language-r">seqlevels(txdb) &lt;- &quot;chr1&quot;
</code></pre>
<p>So if you ran this, then from this point on in your R session only
chromosome 1 would be consulted when you call the various retrieval
methods…  If you need to reset back to the original seqlevels (i.e.
to the seqlevels stored in the db), then set the seqlevels to
<code>seqlevels0(txdb)</code>.</p>
<pre><code class="language-r">seqlevels(txdb) &lt;- seqlevels0(txdb)
</code></pre>
<p>\begin{Exercise} Use </code>seqlevels<code> to set only chromsome 15 to be active.  BTW, the rest of this vignette will assume you have succeeded at this. \end{Exercise}
`5u\begin{Solution}</p>
<pre><code class="language-r">seqlevels(txdb) &lt;- &quot;chr15&quot;
seqlevels(txdb)
</code></pre>
<pre><code>## [1] &quot;chr15&quot;
</code></pre>
<p>\end{Solution}5u`</p>
<h3 id="retrieving-data-using-the-select-method">Retrieving data using the select method</h3>
<p>The <code>TxDb</code> objects inherit from <code>AnnotationDb</code>
objects (just as the <code>ChipDb</code> and <code>OrgDb</code> objects do).
One of the implications of this relationship is that these object
ought to be used in similar ways to each other.  Therefore we have
written supporting <code>columns</code>, <code>keytypes</code>, <code>keys</code>
and <code>select</code> methods for <code>TxDb</code> objects.</p>
<p>These methods can be a useful way of extracting data from a
<code>TxDb</code> object.  And they are used in the same way that
they would be used to extract information about a <code>ChipDb</code> or
an <code>OrgDb</code> object.  Here is a simple example of how to find the
UCSC transcript names that match with a set of gene IDs.</p>
<pre><code class="language-r">keys &lt;- c(&quot;100033416&quot;, &quot;100033417&quot;, &quot;100033420&quot;)
columns(txdb)
</code></pre>
<pre><code>##  [1] &quot;CDSCHROM&quot;   &quot;CDSEND&quot;     &quot;CDSID&quot;      &quot;CDSNAME&quot;    &quot;CDSSTART&quot;  
##  [6] &quot;CDSSTRAND&quot;  &quot;EXONCHROM&quot;  &quot;EXONEND&quot;    &quot;EXONID&quot;     &quot;EXONNAME&quot;  
## [11] &quot;EXONRANK&quot;   &quot;EXONSTART&quot;  &quot;EXONSTRAND&quot; &quot;GENEID&quot;     &quot;TXCHROM&quot;   
## [16] &quot;TXEND&quot;      &quot;TXID&quot;       &quot;TXNAME&quot;     &quot;TXSTART&quot;    &quot;TXSTRAND&quot;  
## [21] &quot;TXTYPE&quot;
</code></pre>
<pre><code class="language-r">keytypes(txdb)
</code></pre>
<pre><code>## [1] &quot;CDSID&quot;    &quot;CDSNAME&quot;  &quot;EXONID&quot;   &quot;EXONNAME&quot; &quot;GENEID&quot;   &quot;TXID&quot;     &quot;TXNAME&quot;
</code></pre>
<pre><code class="language-r">select(txdb, keys = keys, columns=&quot;TXNAME&quot;, keytype=&quot;GENEID&quot;)
</code></pre>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns
</code></pre>
<pre><code>##      GENEID     TXNAME
## 1 100033416 uc001yxl.4
## 2 100033417 uc001yxo.3
## 3 100033420 uc001yxr.3
</code></pre>
<p>\begin{Exercise} For the genes in the example above, find the chromosome and strand information that will go with each of the transcript names. \end{Exercise}
`5u\begin{Solution}</p>
<pre><code class="language-r">columns(txdb)
</code></pre>
<pre><code>##  [1] &quot;CDSCHROM&quot;   &quot;CDSEND&quot;     &quot;CDSID&quot;      &quot;CDSNAME&quot;    &quot;CDSSTART&quot;  
##  [6] &quot;CDSSTRAND&quot;  &quot;EXONCHROM&quot;  &quot;EXONEND&quot;    &quot;EXONID&quot;     &quot;EXONNAME&quot;  
## [11] &quot;EXONRANK&quot;   &quot;EXONSTART&quot;  &quot;EXONSTRAND&quot; &quot;GENEID&quot;     &quot;TXCHROM&quot;   
## [16] &quot;TXEND&quot;      &quot;TXID&quot;       &quot;TXNAME&quot;     &quot;TXSTART&quot;    &quot;TXSTRAND&quot;  
## [21] &quot;TXTYPE&quot;
</code></pre>
<pre><code class="language-r">cols &lt;- c(&quot;TXNAME&quot;, &quot;TXSTRAND&quot;, &quot;TXCHROM&quot;)
select(txdb, keys=keys, columns=cols, keytype=&quot;GENEID&quot;)
</code></pre>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns
</code></pre>
<pre><code>##      GENEID     TXNAME TXCHROM TXSTRAND
## 1 100033416 uc001yxl.4   chr15        +
## 2 100033417 uc001yxo.3   chr15        +
## 3 100033420 uc001yxr.3   chr15        +
</code></pre>
<p>\end{Solution}5u`</p>
<h3 id="methods-for-returning-granges-objects">Methods for returning <code>GRanges</code> objects</h3>
<p>Retrieving data with select is useful, but sometimes it is more
convenient to extract the result as <code>GRanges</code> objects.  This is
often the case when you are doing counting or specialized overlap
operations downstream.  For these use cases there is another family of
methods available.</p>
<p>Perhaps the most common operations for a <code>TxDb</code> object
is to retrieve the genomic coordinates or <em>ranges</em> for exons,
transcripts or coding sequences.  The functions
<code>transcripts</code>, <code>exons</code>, and <code>cds</code> return
the coordinate information as a <code>GRanges</code> object.</p>
<p>As an example, all transcripts present in a <code>TxDb</code> object
can be obtained as follows:</p>
<pre><code class="language-r">GR &lt;- transcripts(txdb)
GR[1:3]
</code></pre>
<pre><code>## GRanges object with 3 ranges and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr15 20362688-20364420      + |     53552  uc001yte.1
##   [2]    chr15 20487997-20496811      + |     53553  uc001ytf.1
##   [3]    chr15 20723929-20727150      + |     53554  uc001ytj.3
##   -------
##   seqinfo: 1 sequence from hg19 genome
</code></pre>
<p>The <code>transcripts</code> function returns a <code>GRanges</code> class
object.  You can learn a lot more about the manipulation of these
objects by reading the <code>GenomicRanges</code> introductory
vignette.  The <code>show</code> method for a <code>GRanges</code> object
will display the ranges, seqnames (a chromosome or a contig), and
strand on the left side and then present related metadata on the right
side.  At the bottom, the seqlengths display all the possible seqnames
along with the length of each sequence.</p>
<p>The <code>strand</code> function is used to obtain the strand
information from the transcripts.  The sum of the Lengths of the
<code>Rle</code> object that <code>strand</code> returns is equal to the
length of the <code>GRanges</code> object.</p>
<pre><code class="language-r">tx_strand &lt;- strand(GR)
tx_strand
</code></pre>
<pre><code>## factor-Rle of length 3337 with 2 runs
##   Lengths: 1732 1605
##   Values :    +    -
## Levels(3): + - *
</code></pre>
<pre><code class="language-r">sum(runLength(tx_strand))
</code></pre>
<pre><code>## [1] 3337
</code></pre>
<pre><code class="language-r">length(GR)
</code></pre>
<pre><code>## [1] 3337
</code></pre>
<p>In addition, the <code>transcripts</code> function can also be used to
retrieve a subset of the transcripts available such as those on the
\(+\)-strand of chromosome 1.</p>
<pre><code class="language-r">GR &lt;- transcripts(txdb, filter=list(tx_chrom = &quot;chr15&quot;, tx_strand = &quot;+&quot;))
length(GR)
</code></pre>
<pre><code>## [1] 1732
</code></pre>
<pre><code class="language-r">unique(strand(GR))
</code></pre>
<pre><code>## [1] +
## Levels: + - *
</code></pre>
<p>The <code>promoters</code> function computes a <code>GRanges</code> object
that spans the promoter region around the transcription start site
for the transcripts in a <code>TxDb</code> object.  The <code>upstream</code>
and <code>downstream</code> arguments define the number of bases upstream
and downstream from the transcription start site that make up the
promoter region.</p>
<pre><code class="language-r">PR &lt;- promoters(txdb, upstream=2000, downstream=400)
PR
</code></pre>
<pre><code>## GRanges object with 3337 ranges and 2 metadata columns:
##              seqnames              ranges strand |     tx_id     tx_name
##                 &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   uc001yte.1    chr15   20360688-20363087      + |     53552  uc001yte.1
##   uc001ytf.1    chr15   20485997-20488396      + |     53553  uc001ytf.1
##   uc001ytj.3    chr15   20721929-20724328      + |     53554  uc001ytj.3
##   uc021sex.1    chr15   20737312-20739711      + |     53555  uc021sex.1
##   uc010tzb.1    chr15   20740235-20742634      + |     53556  uc010tzb.1
##          ...      ...                 ...    ... .       ...         ...
##   uc021syy.1    chr15 102302656-102305055      - |     56884  uc021syy.1
##   uc002cdf.1    chr15 102462863-102465262      - |     56885  uc002cdf.1
##   uc002cds.2    chr15 102518897-102521296      - |     56886  uc002cds.2
##   uc010utv.1    chr15 102518897-102521296      - |     56887  uc010utv.1
##   uc010utw.1    chr15 102518897-102521296      - |     56888  uc010utw.1
##   -------
##   seqinfo: 1 sequence from hg19 genome
</code></pre>
<p>The <code>exons</code> and <code>cds</code> functions can also be used
in a similar fashion to retrive genomic coordinates for exons and
coding sequences.</p>
<p>\begin{Exercise} Use </code>exons<code>to retrieve all the exons from chromosome 15. How does the length of this compare to the value returned by</code>transcripts<code>? \end{Exercise}
`5u\begin{Solution}</p>
<pre><code class="language-r">EX &lt;- exons(txdb)
EX[1:4]
</code></pre>
<pre><code>## GRanges object with 4 ranges and 1 metadata column:
##       seqnames            ranges strand |   exon_id
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   [1]    chr15 20362688-20362858      + |    192986
##   [2]    chr15 20362943-20363123      + |    192987
##   [3]    chr15 20364397-20364420      + |    192988
##   [4]    chr15 20487997-20488227      + |    192989
##   -------
##   seqinfo: 1 sequence from hg19 genome
</code></pre>
<pre><code class="language-r">length(EX)
</code></pre>
<pre><code>## [1] 10771
</code></pre>
<pre><code class="language-r">length(GR)
</code></pre>
<pre><code>## [1] 1732
</code></pre>
<p>\end{Solution}5u`</p>
<h3 id="working-with-grouped-features">Working with Grouped Features</h3>
<p>Often one is interested in how particular genomic features relate to
each other, and not just their location.  For example, it might be of
interest to group transcripts by gene or to group exons by transcript.
Such groupings are supported by the <code>transcriptsBy</code>,
<code>exonsBy</code>, and <code>cdsBy</code> functions.</p>
<p>The following call can be used to group transcripts by genes:</p>
<pre><code class="language-r">GRList &lt;- transcriptsBy(txdb, by = &quot;gene&quot;)
length(GRList)
</code></pre>
<pre><code>## [1] 799
</code></pre>
<pre><code class="language-r">names(GRList)[10:13]
</code></pre>
<pre><code>## [1] &quot;100033424&quot; &quot;100033425&quot; &quot;100033427&quot; &quot;100033428&quot;
</code></pre>
<pre><code class="language-r">GRList[11:12]
</code></pre>
<pre><code>## GRangesList object of length 2:
## $`100033425`
## GRanges object with 1 range and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr15 25324204-25325381      + |     53638  uc001yxw.4
##   -------
##   seqinfo: 1 sequence from hg19 genome
## 
## $`100033427`
## GRanges object with 1 range and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr15 25326433-25326526      + |     53640  uc001yxz.3
##   -------
##   seqinfo: 1 sequence from hg19 genome
</code></pre>
<p>The <code>transcriptsBy</code> function returns a <code>GRangesList</code>
class object.  As with <code>GRanges</code> objects, you can learn more
about these objects by reading the <code>GenomicRanges</code>
introductory vignette.  The <code>show</code> method for a
<code>GRangesList</code> object will display as a list of <code>GRanges</code>
objects.  And, at the bottom the seqinfo will be displayed once for
the entire list.</p>
<p>For each of these three functions, there is a limited set of options
that can be passed into the <code>by</code> argument to allow grouping.
For the <code>transcriptsBy</code> function, you can group by gene,
exon or cds, whereas for the <code>exonsBy</code> and <code>cdsBy</code>
functions can only be grouped by transcript (tx) or gene.</p>
<p>So as a further example, to extract all the exons for each transcript
you can call:</p>
<pre><code class="language-r">GRList &lt;- exonsBy(txdb, by = &quot;tx&quot;)
length(GRList)
</code></pre>
<pre><code>## [1] 3337
</code></pre>
<pre><code class="language-r">names(GRList)[10:13]
</code></pre>
<pre><code>## [1] &quot;53561&quot; &quot;53562&quot; &quot;53563&quot; &quot;53564&quot;
</code></pre>
<pre><code class="language-r">GRList[[12]]
</code></pre>
<pre><code>## GRanges object with 1 range and 3 metadata columns:
##       seqnames            ranges strand |   exon_id   exon_name exon_rank
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
##   [1]    chr15 22043463-22043502      + |    193028        &lt;NA&gt;         1
##   -------
##   seqinfo: 1 sequence from hg19 genome
</code></pre>
<p>As you can see, the <code>GRangesList</code> objects returned from each
function contain locations and identifiers grouped into a list like
object according to the type of feature specified in the <code>by</code>
argument. The object returned can then be used by functions like
<code>findOverlaps</code> to contextualize alignments from
high-throughput sequencing.</p>
<p>The identifiers used to label the <code>GRanges</code> objects depend upon
the data source used to create the <code>TxDb</code> object.  So
the list identifiers will not always be Entrez Gene IDs, as they were
in the first example.  Furthermore, some data sources do not provide a
unique identifier for all features.  In this situation, the group
label will be a synthetic ID created by <code>GenomicFeatures</code> to
keep the relations between features consistent in the database this
was the case in the 2nd example.  Even though the results will
sometimes have to come back to you as synthetic IDs, you can still
always retrieve the original IDs.</p>
<p>\begin{Exercise} Starting with the tx_ids that are the names of the GRList object we just made, use </code>select<code>to retrieve that matching transcript names.  Remember that the list used a</code>by<code> argument = &quot;tx&quot;, so the list is grouped by transcript IDs. \end{Exercise}
`5u\begin{Solution}</p>
<pre><code class="language-r">GRList &lt;- exonsBy(txdb, by = &quot;tx&quot;)
tx_ids &lt;- names(GRList)
head(select(txdb, keys=tx_ids, columns=&quot;TXNAME&quot;, keytype=&quot;TXID&quot;))
</code></pre>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns
</code></pre>
<pre><code>##    TXID     TXNAME
## 1 53552 uc001yte.1
## 2 53553 uc001ytf.1
## 3 53554 uc001ytj.3
## 4 53555 uc021sex.1
## 5 53556 uc010tzb.1
## 6 53557 uc021sey.1
</code></pre>
<p>\end{Solution}5u`</p>
<p>Finally, the order of the results in a <code>GRangesList</code> object can
vary with the way in which things were grouped. In most cases the
grouped elements of the <code>GRangesList</code> object will be listed in
the order that they occurred along the chromosome.  However, when
exons or CDS parts are grouped by transcript, they will instead be
grouped according to their position along the transcript itself.
This is important because alternative splicing can mean that the
order along the transcript can be different from that along the
chromosome.</p>
<h3 id="predefined-grouping-functions">Predefined grouping functions</h3>
<p>The <code>intronsByTranscript</code>, <code>fiveUTRsByTranscript</code>
and <code>threeUTRsByTranscript</code> are convenience functions that
provide behavior equivalent to the grouping functions, but in
prespecified form. These functions return a <code>GRangesList</code>
object grouped by transcript for introns, 5’ UTR’s, and 3’ UTR’s,
respectively.  Below are examples of how you can call these methods.</p>
<pre><code class="language-r">length(intronsByTranscript(txdb))
</code></pre>
<pre><code>## [1] 3337
</code></pre>
<pre><code class="language-r">length(fiveUTRsByTranscript(txdb))
</code></pre>
<pre><code>## [1] 1825
</code></pre>
<pre><code class="language-r">length(threeUTRsByTranscript(txdb))
</code></pre>
<pre><code>## [1] 1803
</code></pre>
<h3 id="getting-the-actual-sequence-data">Getting the actual sequence data</h3>
<p>The <code>GenomicFeatures</code> package also provides provides
functions for converting from ranges to actual sequence (when paired
with an appropriate <code>BSgenome</code> package).</p>
<pre><code class="language-r">library(BSgenome.Hsapiens.UCSC.hg19)
</code></pre>
<pre><code>## Loading required package: BSgenome
</code></pre>
<pre><code>## Loading required package: Biostrings
</code></pre>
<pre><code>## Loading required package: XVector
</code></pre>
<pre><code>## 
## Attaching package: 'Biostrings'
</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     strsplit
</code></pre>
<pre><code>## Loading required package: BiocIO
</code></pre>
<pre><code>## Loading required package: rtracklayer
</code></pre>
<pre><code>## 
## Attaching package: 'rtracklayer'
</code></pre>
<pre><code>## The following object is masked from 'package:BiocIO':
## 
##     FileForFormat
</code></pre>
<pre><code class="language-r">tx_seqs1 &lt;- extractTranscriptSeqs(Hsapiens, TxDb.Hsapiens.UCSC.hg19.knownGene,
                                  use.names=TRUE)
</code></pre>
<p>And, once these sequences have been extracted, you can translate them
into proteins with <code>translate</code>:</p>
<pre><code class="language-r">suppressWarnings(translate(tx_seqs1))
</code></pre>
<pre><code>## AAStringSet object of length 3337:
##        width seq                                            names               
##    [1]   125 EDQDDEARVQYEGFRPGMYVRV...YTPQHMHCGAAFWA*FSDSCH uc001yte.1
##    [2]   288 RIAS*GRAEFSSAQTSEIQRRR...ESVFYSVYFNYGNNCFFTVTD uc001ytf.1
##    [3]   588 RSGQRLPEQPEAEGGDPGKQRR...RDLLENETHLYLCSIKICFSS uc001ytj.3
##    [4]    10 HHLNCRPQTG                                     uc021sex.1
##    [5]     9 STVTLPHSQ                                      uc010tzb.1
##    ...   ... ...
## [3333]    10 QVPMRVQVGQ                                     uc021syy.1
## [3334]   306 MVTEFIFLGLSDSQELQTFLFM...DMKTAIRRLRKWDAHSSVKF* uc002cdf.1
## [3335]   550 LAVSLFFDLFFLFMCICCLLAQ...TPRRLHPAQLEILY*KHTVGF uc002cds.2
## [3336]   496 LAVSLFFDLFFLFMCICCLLAQ...PETFASCTARDPLLKAHCWFL uc010utv.1
## [3337]   531 LAVSLFFDLFFLFMCICCLLAQ...TPRRLHPAQLEILY*KHTVGF uc010utw.1
</code></pre>
<p>\begin{Exercise} But of course this is not a meaningful translation, because the call to </code>extractTranscriptSeqs<code>will have extracted all the transcribed regions of the genome regardless of whether or not they are translated. Look at the manual page for</code>extractTranscriptSeqs<code> and see how you can use cdsBy to only translate only the coding regions. \end{Exercise}
`5u\begin{Solution}</p>
<pre><code class="language-r">cds_seqs &lt;- extractTranscriptSeqs(Hsapiens,
                                  cdsBy(txdb, by=&quot;tx&quot;, use.names=TRUE))
translate(cds_seqs)
</code></pre>
<pre><code>## AAStringSet object of length 1875:
##        width seq                                            names               
##    [1]   102 MYVRVEIENVPCEFVQNIDPHY...RQRLLKYTPQHMHCGAAFWA* uc001yte.1
##    [2]   435 MEWKLEQSMREQALLKAQLTQL...LGSNCCVPFFCWAWPPRRRR* uc010tzc.1
##    [3]   317 MKIANNTVVTEFILLGLTQSQD...SMKRLLSRHVVCQVDFIIRN* uc001yuc.1
##    [4]   314 METANYTKVTEFVLTGLSQTPE...KEVKAAMRKLVTKYILCKEK* uc010tzu.2
##    [5]   317 MKIANNTVVTEFILLGLTQSQD...SMKRLLSRHVVCQVDFIIRN* uc010tzv.2
##    ...   ... ...
## [1871]   186 MAGGVLPLRGLRALCRVLLFLS...CLGRSEFKDICQQNVFLQVY* uc010ush.1
## [1872]   258 MYNSKLWEASGHWQHYSENMFT...PVNFLKKDLWLTLTWITVVH* uc002bxl.3
## [1873]   803 MAAEALAAEAVASRLERQEEDI...AIDKLKNLRKTRTLNAEEAF* uc002bxm.3
## [1874]   306 MVTEFIFLGLSDSQELQTFLFM...DMKTAIRRLRKWDAHSSVKF* uc002cdf.1
## [1875]   134 MSESINFSHNLGQLLSPPRCVV...KGETQESVESRVLPGPRHRH* uc010utv.1
</code></pre>
<p>\end{Solution}5u`</p>
<h2 id="creating-new-txdb-objects-or-packages">Creating New <code>TxDb</code> Objects or Packages</h2>
<p>The <code>GenomicFeatures</code> package provides functions to create
<code>TxDb</code> objects based on data downloaded from UCSC
Genome Bioinformatics or BioMart. The following subsections
demonstrate the use of these functions.  There is also support for
creating <code>TxDb</code> objects from custom data sources using
<code>makeTxDb</code>; see the help page for this function for
details.</p>
<h3 id="using-maketxdbfromucsc">Using <code>makeTxDbFromUCSC</code></h3>
<p>The function <code>makeTxDbFromUCSC</code> downloads UCSC
Genome Bioinformatics transcript tables (e.g. <code>knownGene</code>,
<code>refGene</code>, <code>ensGene</code>) for a genome build (e.g.
<code>mm9</code>, <code>hg19</code>).  Use the <code>supportedUCSCtables</code>
utility function to get the list of tables known to work with
<code>makeTxDbFromUCSC</code>.</p>
<pre><code class="language-r">supportedUCSCtables(genome=&quot;mm9&quot;)
</code></pre>
<pre><code>##         tablename              track composite_track
## 1         acembly      AceView Genes            &lt;NA&gt;
## 2    augustusGene           AUGUSTUS            &lt;NA&gt;
## 3        ccdsGene               CCDS            &lt;NA&gt;
## 4         ensGene      Ensembl Genes            &lt;NA&gt;
## 5        exoniphy           Exoniphy            &lt;NA&gt;
## 6          geneid       Geneid Genes            &lt;NA&gt;
## 7         genscan      Genscan Genes            &lt;NA&gt;
## 8       knownGene         UCSC Genes            &lt;NA&gt;
## 9   knownGeneOld4     Old UCSC Genes            &lt;NA&gt;
## 10      nscanGene             N-SCAN            &lt;NA&gt;
## 11   pseudoYale60      Yale Pseudo60            &lt;NA&gt;
## 12        refGene        UCSC RefSeq     NCBI RefSeq
## 13        sgpGene          SGP Genes            &lt;NA&gt;
## 14  transcriptome      Transcriptome            &lt;NA&gt;
## 15 vegaPseudoGene   Vega Pseudogenes      Vega Genes
## 16       vegaGene Vega Protein Genes      Vega Genes
## 17    xenoRefGene       Other RefSeq            &lt;NA&gt;
</code></pre>
<pre><code class="language-r">mm9KG_txdb &lt;- makeTxDbFromUCSC(genome=&quot;mm9&quot;, tablename=&quot;knownGene&quot;)
</code></pre>
<h3 id="using-maketxdbfrombiomart">Using <code>makeTxDbFromBiomart</code></h3>
<p>Retrieve data from BioMart by specifying the mart and the data set to
the <code>makeTxDbFromBiomart</code> function (not all BioMart
data sets are currently supported):</p>
<pre><code class="language-r">mmusculusEnsembl &lt;- makeTxDbFromBiomart(dataset=&quot;mmusculus_gene_ensembl&quot;)
</code></pre>
<p>As with the <code>makeTxDbFromUCSC</code> function, the
<code>makeTxDbFromBiomart</code> function also has a
<code>circ_seqs</code> argument that will default to using the contents
of the <code>DEFAULT_CIRC_SEQS</code> vector.  And just like those UCSC
sources, there is also a helper function called
<code>getChromInfoFromBiomart</code> that can show what the different
chromosomes are called for a given source.</p>
<p>Using the <code>makeTxDbFromBiomart</code>
<code>makeTxDbFromUCSC</code> functions can take a while and
may also require some bandwidth as these methods have to download and
then assemble a database from their respective sources.  It is not
expected that most users will want to do this step every time.
Instead, we suggest that you save your annotation objects and label
them with an appropriate time stamp so as to facilitate reproducible
research.</p>
<h3 id="using-maketxdbfromensembl">Using <code>makeTxDbFromEnsembl</code></h3>
<p>The <code>makeTxDbFromEnsembl</code> function creates a <code>TxDb</code> object
for a given organism by importing the genomic locations of its transcripts,
exons, CDS, and genes from an Ensembl database.</p>
<p>See <code>?makeTxDbFromEnsembl</code> for more information.</p>
<h3 id="using-maketxdbfromgff">Using <code>makeTxDbFromGFF</code></h3>
<p>You can also extract transcript information from either GFF3 or GTF
files by using the <code>makeTxDbFromGFF</code> function.
Usage is similar to <code>makeTxDbFromBiomart</code> and
<code>makeTxDbFromUCSC</code>.</p>
<h3 id="saving-and-loading-a-txdb-object">Saving and Loading a <code>TxDb</code> Object</h3>
<p>Once a <code>TxDb</code> object has been created, it can be saved
to avoid the time and bandwidth costs of recreating it and to make it
possible to reproduce results with identical genomic feature data at a
later date.  Since <code>TxDb</code> objects are backed by a
SQLite database, the save format is a SQLite database file (which
could be accessed from programs other than R if desired).  Note that
it is not possible to serialize a <code>TxDb</code> object using
R’s <code>save</code> function.</p>
<pre><code class="language-r">saveDb(mm9KG_txdb, file=&quot;fileName.sqlite&quot;)
</code></pre>
<p>And as was mentioned earlier, a saved <code>TxDb</code> object can
be initialized from a .sqlite file by simply using <code>loadDb</code>.</p>
<pre><code class="language-r">mm9KG_txdb &lt;- loadDb(&quot;fileName.sqlite&quot;)
</code></pre>
<h3 id="using-maketxdbpackagefromucsc-and">Using <code>makeTxDbPackageFromUCSC</code> and</h3>
<pre><code>`makeTxDbPackageFromBiomart`
</code></pre>
<p>It is often much more convenient to just make an annotation package
out of your annotations.  If you are finding that this is the case,
then you should consider the convenience functions:
<code>makeTxDbPackageFromUCSC</code> and
<code>makeTxDbPackageFromBiomart</code>.  These functions are similar
to <code>makeTxDbFromUCSC</code> and
<code>makeTxDbFromBiomart</code> except that they will take the
extra step of actually wrapping the database up into an annotation
package for you.  This package can then be installed and used as of
the standard TxDb packages found on in the Bioconductor
repository.</p>
<h2 id="session-information">Session Information</h2>
<pre><code>## R version 4.3.3 (2024-02-29)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 22.04.4 LTS
## 
## Matrix products: default
## BLAS:   /home/biocbuild/bbs-3.18-bioc/R/lib/libRblas.so 
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB              LC_COLLATE=C              
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: America/New_York
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] BSgenome.Hsapiens.UCSC.hg19_1.4.3      
##  [2] BSgenome_1.70.2                        
##  [3] rtracklayer_1.62.0                     
##  [4] BiocIO_1.12.0                          
##  [5] Biostrings_2.70.3                      
##  [6] XVector_0.42.0                         
##  [7] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2
##  [8] GenomicFeatures_1.54.4                 
##  [9] AnnotationDbi_1.64.1                   
## [10] Biobase_2.62.0                         
## [11] GenomicRanges_1.54.1                   
## [12] GenomeInfoDb_1.38.7                    
## [13] IRanges_2.36.0                         
## [14] S4Vectors_0.40.2                       
## [15] BiocGenerics_0.48.1                    
## 
## loaded via a namespace (and not attached):
##  [1] KEGGREST_1.42.0             SummarizedExperiment_1.32.0
##  [3] rjson_0.2.21                xfun_0.42                  
##  [5] lattice_0.22-5              vctrs_0.6.5                
##  [7] tools_4.3.3                 bitops_1.0-7               
##  [9] generics_0.1.3              curl_5.2.1                 
## [11] parallel_4.3.3              tibble_3.2.1               
## [13] fansi_1.0.6                 RSQLite_2.3.5              
## [15] blob_1.2.4                  pkgconfig_2.0.3            
## [17] Matrix_1.6-5                dbplyr_2.4.0               
## [19] lifecycle_1.0.4             GenomeInfoDbData_1.2.11    
## [21] compiler_4.3.3              stringr_1.5.1              
## [23] Rsamtools_2.18.0            progress_1.2.3             
## [25] codetools_0.2-19            RCurl_1.98-1.14            
## [27] yaml_2.3.8                  pillar_1.9.0               
## [29] crayon_1.5.2                BiocParallel_1.36.0        
## [31] DelayedArray_0.28.0         cachem_1.0.8               
## [33] abind_1.4-5                 tidyselect_1.2.1           
## [35] digest_0.6.35               stringi_1.8.3              
## [37] dplyr_1.1.4                 restfulr_0.0.15            
## [39] grid_4.3.3                  biomaRt_2.58.2             
## [41] fastmap_1.1.1               SparseArray_1.2.4          
## [43] cli_3.6.2                   magrittr_2.0.3             
## [45] S4Arrays_1.2.1              XML_3.99-0.16.1            
## [47] utf8_1.2.4                  prettyunits_1.2.0          
## [49] filelock_1.0.3              rappdirs_0.3.3             
## [51] bit64_4.0.5                 httr_1.4.7                 
## [53] matrixStats_1.2.0           bit_4.0.5                  
## [55] png_0.1-8                   hms_1.1.3                  
## [57] memoise_2.0.1               evaluate_0.23              
## [59] knitr_1.45                  BiocFileCache_2.10.1       
## [61] rlang_1.1.3                 glue_1.7.0                 
## [63] DBI_1.2.2                   xml2_1.3.6                 
## [65] R6_2.5.1                    MatrixGenerics_1.14.0      
## [67] GenomicAlignments_1.38.2    zlibbioc_1.48.2
</code></pre>
</div>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/components/prism-core.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/plugins/autoloader/prism-autoloader.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/combine/npm/katex/dist/katex.min.js,npm/katex/dist/contrib/auto-render.min.js,npm/@xiee/utils/js/render-katex.js" defer></script>
</body>
</html>
