<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>KEGGREST</title>
<style type="text/css">
/**
 * Prism.s theme ported from highlight.js's xcode style
 */
pre code {
  padding: 1em;
}
.token.comment {
  color: #007400;
}
.token.punctuation {
  color: #999;
}
.token.tag,
.token.selector {
  color: #aa0d91;
}
.token.boolean,
.token.number,
.token.constant,
.token.symbol {
  color: #1c00cf;
}
.token.property,
.token.attr-name,
.token.string,
.token.char,
.token.builtin {
  color: #c41a16;
}
.token.inserted {
  background-color: #ccffd8;
}
.token.deleted {
  background-color: #ffebe9;
}
.token.operator,
.token.entity,
.token.url,
.language-css .token.string,
.style .token.string {
  color: #9a6e3a;
}
.token.atrule,
.token.attr-value,
.token.keyword {
  color: #836c28;
}
.token.function,
.token.class-name {
  color: #DD4A68;
}
.token.regex,
.token.important,
.token.variable {
  color: #5c2699;
}
.token.important,
.token.bold {
  font-weight: bold;
}
.token.italic {
  font-style: italic;
}
</style>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  box-sizing: border-box;
}
body, .footnotes, code { font-size: .9em; }
li li { font-size: .95em; }
*, *:before, *:after {
  box-sizing: inherit;
}
pre, img { max-width: 100%; }
pre, pre:hover {
  white-space: pre-wrap;
  word-break: break-all;
}
pre code {
  display: block;
  overflow-x: auto;
}
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre) > code, code[class] { background-color: #F8F8F8; }
code.language-undefined, pre > code:not([class]) {
  background-color: inherit;
  border: 1px solid #eee;
}
table {
  margin: auto;
  border-top: 1px solid #666;
}
table thead th { border-bottom: 1px solid #ddd; }
th, td { padding: 5px; }
thead, tfoot, tr:nth-child(even) { background: #eee; }
blockquote {
  color: #666;
  margin: 0;
  padding-left: 1em;
  border-left: 0.5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC .numbered li { list-style: none; }
#TOC .numbered { padding-left: 0; }
#TOC .numbered ul { padding-left: 1em; }
table, .body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.footnote-ref a::before { content: "["; }
.footnote-ref a::after { content: "]"; }
section.footnotes::before {
  content: "";
  display: block;
  max-width: 20em;
}

@media print {
  body {
    font-size: 12pt;
    max-width: 100%;
  }
  tr, img { page-break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  pre { white-space: pre; }
}
</style>
</head>
<body>
<div class="include-before">
</div>
<div class="frontmatter">
<div class="title"><h1>KEGGREST</h1></div>
<div class="author"><h2></h2></div>
<div class="date"><h3></h3></div>
</div>
<div class="body">
<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Accessing the KEGG REST API}
-->
<h1 id="keggrest">KEGGREST</h1>
<p><a href="https://www.kegg.jp/kegg/">KEGG</a>
is a database resource for understanding high-level functions
and utilities of the biological system, such as the cell, the organism
and the ecosystem, from molecular-level information, especially
large-scale molecular datasets generated by genome sequencing and
other high-throughput experimental technologies.</p>
<p><code>KEGGREST</code> allows access to the
<a href="https://www.kegg.jp/kegg/docs/keggapi.html">KEGG REST API</a>. Since
KEGG disabled the KEGG SOAP server
on December 31, 2012 (which means the <code>KEGGSOAP</code> package will no
longer work), <code>KEGGREST</code> serves as a replacement.</p>
<p>The interface to <code>KEGGREST</code> is simpler and in some ways more
powerful than <code>KEGGSOAP</code>; however, not all the functionality
that was available through the SOAP API has been exposed
in the REST API. If and when more functionality is exposed
on the server side, this package will be updated to take
advantage of it.</p>
<h2 id="overview">Overview</h2>
<p>The KEGG REST API is built on some simple operations:
<code>info</code>, <code>list</code>, <code>find</code>, <code>get</code>, <code>conv</code>, and <code>link</code>.
The corresponding <code>R</code> functions in <code>KEGGREST</code> are:
<code>keggInfo()</code>, <code>keggList()</code>, <code>keggFind()</code>, <code>keggGet()</code>,
<code>keggConv</code>, and <code>keggLink()</code>.</p>
<h2 id="exploring-kegg-resources-with-kegglist">Exploring KEGG Resources with <code>keggList()</code></h2>
<p>KEGG exposes a number of databases. To get an idea of
what is available, run <code>listDatabases()</code>:</p>
<pre><code class="language-r">library(KEGGREST)
listDatabases()
</code></pre>
<pre><code>##  [1] &quot;pathway&quot;  &quot;brite&quot;    &quot;module&quot;   &quot;ko&quot;       &quot;genome&quot;   &quot;vg&quot;      
##  [7] &quot;ag&quot;       &quot;compound&quot; &quot;glycan&quot;   &quot;reaction&quot; &quot;rclass&quot;   &quot;enzyme&quot;  
## [13] &quot;disease&quot;  &quot;drug&quot;     &quot;dgroup&quot;   &quot;environ&quot;  &quot;genes&quot;    &quot;ligand&quot;  
## [19] &quot;kegg&quot;
</code></pre>
<p>You can use these databases in further queries. Note that in many
cases you can also use a three-letter KEGG organism code or a
“T number” (genome identifier) in the same place you would use
one of these database names.</p>
<p>You can obtain the list of organisms available in KEGG with
the <code>keggList()</code> function:</p>
<pre><code class="language-r">org &lt;- keggList(&quot;organism&quot;)
head(org)
</code></pre>
<pre><code>##      T.number organism species                                              
## [1,] &quot;T01001&quot; &quot;hsa&quot;    &quot;Homo sapiens (human)&quot;                               
## [2,] &quot;T01005&quot; &quot;ptr&quot;    &quot;Pan troglodytes (chimpanzee)&quot;                       
## [3,] &quot;T02283&quot; &quot;pps&quot;    &quot;Pan paniscus (bonobo)&quot;                              
## [4,] &quot;T02442&quot; &quot;ggo&quot;    &quot;Gorilla gorilla gorilla (western lowland gorilla)&quot;  
## [5,] &quot;T01416&quot; &quot;pon&quot;    &quot;Pongo abelii (Sumatran orangutan)&quot;                  
## [6,] &quot;T03265&quot; &quot;nle&quot;    &quot;Nomascus leucogenys (northern white-cheeked gibbon)&quot;
##      phylogeny                               
## [1,] &quot;Eukaryotes;Animals;Vertebrates;Mammals&quot;
## [2,] &quot;Eukaryotes;Animals;Vertebrates;Mammals&quot;
## [3,] &quot;Eukaryotes;Animals;Vertebrates;Mammals&quot;
## [4,] &quot;Eukaryotes;Animals;Vertebrates;Mammals&quot;
## [5,] &quot;Eukaryotes;Animals;Vertebrates;Mammals&quot;
## [6,] &quot;Eukaryotes;Animals;Vertebrates;Mammals&quot;
</code></pre>
<p>From <code>KEGGREST</code>‘s point of view, you’ve just asked KEGG
to show you the name of every entry in the “organism” database.</p>
<p>Therefore, the complete list of entities that can be
queried with <code>KEGGREST</code> can be obtained as follows:</p>
<pre><code class="language-r">queryables &lt;- c(listDatabases(), org[,1], org[,2])
</code></pre>
<p>You could also ask for every entry in the “hsa” (<em>Homo sapiens</em>)
database as follows:</p>
<pre><code class="language-r">keggList(&quot;hsa&quot;)
</code></pre>
<h2 id="get-specific-entries-with-keggget">Get specific entries with <code>keggGet()</code></h2>
<p>Once you have a list of specific KEGG identifiers, use
<code>keggGet()</code> to get more information about them. Here we look up
a human gene and an E. coli O157 gene:</p>
<pre><code class="language-r">query &lt;- keggGet(c(&quot;hsa:10458&quot;, &quot;ece:Z5100&quot;))
</code></pre>
<p>As expected, this returns two items:</p>
<pre><code class="language-r">length(query)
</code></pre>
<pre><code>## [1] 2
</code></pre>
<p>Behind the scenes, <code>KEGGREST</code> downloaded and parsed a KEGG
<a href="https://www.kegg.jp/kegg/rest/dbentry.html">flat file</a>, which you
can now explore:</p>
<pre><code class="language-r">names(query[[1]])
</code></pre>
<pre><code>##  [1] &quot;ENTRY&quot;     &quot;SYMBOL&quot;    &quot;NAME&quot;      &quot;ORTHOLOGY&quot; &quot;ORGANISM&quot;  &quot;PATHWAY&quot;  
##  [7] &quot;NETWORK&quot;   &quot;BRITE&quot;     &quot;POSITION&quot;  &quot;MOTIF&quot;     &quot;DBLINKS&quot;   &quot;STRUCTURE&quot;
## [13] &quot;AASEQ&quot;     &quot;NTSEQ&quot;
</code></pre>
<pre><code class="language-r">query[[1]]$ENTRY
</code></pre>
<pre><code>##     CDS 
## &quot;10458&quot;
</code></pre>
<pre><code class="language-r">query[[1]]$DBLINKS
</code></pre>
<pre><code>## [1] &quot;NCBI-GeneID: 10458&quot;        &quot;NCBI-ProteinID: NP_059345&quot;
## [3] &quot;OMIM: 605475&quot;              &quot;HGNC: 947&quot;                
## [5] &quot;Ensembl: ENSG00000175866&quot;  &quot;UniProt: Q9UQB8&quot;
</code></pre>
<p><code>keggGet()</code> can also return amino acid sequences as <code>AAStringSet</code> objects
(from the <code>Biostrings</code> package):</p>
<pre><code class="language-r">keggGet(c(&quot;hsa:10458&quot;, &quot;ece:Z5100&quot;), &quot;aaseq&quot;) ## retrieves amino acid sequences
</code></pre>
<pre><code>## AAStringSet object of length 2:
##     width seq                                               names               
## [1]   552 MSLSRSEEMHRLTENVYKTIMEQ...DLSAQGPEGREHGDGSARTLAGR hsa:10458 K05627 ...
## [2]   248 MLNGISNAASTLGRQLVGIASRV...SGLPPLAQALKDHLAAYEQSKKG ece:Z5100 K12786 ...
</code></pre>
<p>…or <code>DNAStringSet</code> objects if <code>option</code> is <code>ntseq</code>:</p>
<pre><code class="language-r">keggGet(c(&quot;hsa:10458&quot;, &quot;ece:Z5100&quot;), &quot;ntseq&quot;) ## retrieves nucleotide sequences
</code></pre>
<pre><code>## DNAStringSet object of length 2:
##     width seq                                               names               
## [1]  1659 ATGTCTCTGTCTCGCTCAGAGGA...CCCGCACCCTGGCTGGAAGATGA hsa:10458 K05627 ...
## [2]   747 ATGCTTAATGGAATTAGTAACGC...ATGAGCAATCGAAGAAAGGGTAA ece:Z5100 K12786 ...
</code></pre>
<p><code>keggGet()</code> can also return images:</p>
<pre><code class="language-r">png &lt;- keggGet(&quot;hsa05130&quot;, &quot;image&quot;) 
t &lt;- tempfile()
library(png)
writePNG(png, t)
if (interactive()) browseURL(t)
</code></pre>
<p><strong>NOTE</strong>: <code>keggGet()</code> can only return 10 result sets at once (this limitation
is on the server side). If you supply more than 10 inputs to <code>keggGet()</code>,
<code>KEGGREST</code> will warn that only the first 10 results will be returned.</p>
<h2 id="search-by-keywords-with-keggfind">Search by keywords with <code>keggFind()</code></h2>
<p>You can search for two separate keywords (“shiga” and “toxin” in this case):</p>
<pre><code class="language-r">head(keggFind(&quot;genes&quot;, c(&quot;shiga&quot;, &quot;toxin&quot;)))
</code></pre>
<pre><code>##                                                               ece:Z1464 
## &quot;stx2A; shiga-like toxin II A subunit encoded by bacteriophage BP-933W&quot; 
##                                                               ece:Z1465 
## &quot;stx2B; shiga-like toxin II B subunit encoded by bacteriophage BP-933W&quot; 
##                                                               ece:Z3343 
##   &quot;stx1B; shiga-like toxin 1 subunit B encoded within prophage CP-933V&quot; 
##                                                               ece:Z3344 
##   &quot;stx1A; shiga-like toxin 1 subunit A encoded within prophage CP-933V&quot; 
##                                                            ecs:ECs_1205 
##                                        &quot;stx2A; Shiga toxin 2 subunit A&quot; 
##                                                            ecs:ECs_1206 
##                                        &quot;stx2B; Shiga toxin 2 subunit B&quot;
</code></pre>
<p>Or search for the two words together:</p>
<pre><code class="language-r">head(keggFind(&quot;genes&quot;, &quot;shiga toxin&quot;))
</code></pre>
<pre><code>##                                                               ece:Z1464 
## &quot;stx2A; shiga-like toxin II A subunit encoded by bacteriophage BP-933W&quot; 
##                                                               ece:Z1465 
## &quot;stx2B; shiga-like toxin II B subunit encoded by bacteriophage BP-933W&quot; 
##                                                               ece:Z3343 
##   &quot;stx1B; shiga-like toxin 1 subunit B encoded within prophage CP-933V&quot; 
##                                                               ece:Z3344 
##   &quot;stx1A; shiga-like toxin 1 subunit A encoded within prophage CP-933V&quot; 
##                                                            ecs:ECs_1205 
##                                        &quot;stx2A; Shiga toxin 2 subunit A&quot; 
##                                                            ecs:ECs_1206 
##                                        &quot;stx2B; Shiga toxin 2 subunit B&quot;
</code></pre>
<p>Search for a chemical formula:</p>
<pre><code class="language-r">head(keggFind(&quot;compound&quot;, &quot;C7H10O5&quot;, &quot;formula&quot;))
</code></pre>
<pre><code>## cpd:C00493 cpd:C04236 cpd:C16588 cpd:C17696 cpd:C18307 cpd:C18312 
##  &quot;C7H10O5&quot;  &quot;C7H10O5&quot;  &quot;C7H10O5&quot;  &quot;C7H10O5&quot;  &quot;C7H10O5&quot;  &quot;C7H10O5&quot;
</code></pre>
<p>Search for a chemical formula containing “O5” and “C7”:</p>
<pre><code class="language-r">head(keggFind(&quot;compound&quot;, &quot;O5C7&quot;, &quot;formula&quot;))
</code></pre>
<pre><code>## cpd:C00493 cpd:C00624 cpd:C01215 cpd:C01424 cpd:C02123 cpd:C02236 
##  &quot;C7H10O5&quot; &quot;C7H11NO5&quot;  &quot;C7H9NO5&quot;   &quot;C7H6O5&quot;  &quot;C7H12O5&quot;  &quot;C7H6O5S&quot;
</code></pre>
<p>You can search for compounds with a particular exact mass:</p>
<pre><code class="language-r">keggFind(&quot;compound&quot;, 174.05, &quot;exact_mass&quot;)
</code></pre>
<pre><code>##   cpd:C00493   cpd:C04236   cpd:C16588   cpd:C17696   cpd:C18307   cpd:C18312 
## &quot;174.052823&quot; &quot;174.052823&quot; &quot;174.052823&quot; &quot;174.052823&quot; &quot;174.052823&quot; &quot;174.052823&quot; 
##   cpd:C21281 
## &quot;174.052823&quot;
</code></pre>
<p>Because we’ve supplied a number with two decimal digits of precision,
KEGG will find all compounds with exact mass between 174.045 and 174.055.</p>
<p>Integer ranges can be used to find compounds by molecular weight:</p>
<pre><code class="language-r">head(keggFind(&quot;compound&quot;, 300:310, &quot;mol_weight&quot;))
</code></pre>
<pre><code>##   cpd:C00051   cpd:C00200   cpd:C00219   cpd:C00239   cpd:C00270   cpd:C00357 
##  &quot;307.32348&quot;  &quot;306.33696&quot;  &quot;304.46688&quot; &quot;307.197122&quot;  &quot;309.26986&quot; &quot;301.187702&quot;
</code></pre>
<h2 id="convert-identifiers-with-keggconv">Convert identifiers with <code>keggConv()</code></h2>
<p>Convert between KEGG identifiers and outside identifiers.</p>
<p>You can either specify fully qualified identifiers:</p>
<pre><code class="language-r">keggConv(&quot;ncbi-proteinid&quot;, c(&quot;hsa:10458&quot;, &quot;ece:Z5100&quot;))
</code></pre>
<pre><code>##                  hsa:10458                  ece:Z5100 
## &quot;ncbi-proteinid:NP_059345&quot;  &quot;ncbi-proteinid:AAG58814&quot;
</code></pre>
<p>…or get the mapping for an entire species:</p>
<pre><code class="language-r">head(keggConv(&quot;eco&quot;, &quot;ncbi-geneid&quot;))
</code></pre>
<pre><code>## ncbi-geneid:944742 ncbi-geneid:945803 ncbi-geneid:947498 ncbi-geneid:945198 
##        &quot;eco:b0001&quot;        &quot;eco:b0002&quot;        &quot;eco:b0003&quot;        &quot;eco:b0004&quot; 
## ncbi-geneid:944747 ncbi-geneid:944749 
##        &quot;eco:b0005&quot;        &quot;eco:b0006&quot;
</code></pre>
<p>Reversing the arguments does the opposite mapping:</p>
<pre><code class="language-r">head(keggConv(&quot;ncbi-geneid&quot;, &quot;eco&quot;))
</code></pre>
<pre><code>##            eco:b0001            eco:b0002            eco:b0003 
## &quot;ncbi-geneid:944742&quot; &quot;ncbi-geneid:945803&quot; &quot;ncbi-geneid:947498&quot; 
##            eco:b0004            eco:b0005            eco:b0006 
## &quot;ncbi-geneid:945198&quot; &quot;ncbi-geneid:944747&quot; &quot;ncbi-geneid:944749&quot;
</code></pre>
<h2 id="link-across-databases-with-kegglink">Link across databases with <code>keggLink()</code></h2>
<p>Most of the <code>KEGGSOAP</code> functions whose names started with
“get”, for example <code>get.pathways.by.genes()</code>, can be replaced
with the <code>keggLink()</code> function. Here we query all pathways
for human:</p>
<pre><code class="language-r">head(keggLink(&quot;pathway&quot;, &quot;hsa&quot;))
</code></pre>
<pre><code>##       hsa:10327         hsa:124         hsa:125         hsa:126         hsa:127 
## &quot;path:hsa00010&quot; &quot;path:hsa00010&quot; &quot;path:hsa00010&quot; &quot;path:hsa00010&quot; &quot;path:hsa00010&quot; 
##         hsa:128 
## &quot;path:hsa00010&quot;
</code></pre>
<p>…but you can also specify one or more genes (from multiple species):</p>
<pre><code class="language-r">keggLink(&quot;pathway&quot;, c(&quot;hsa:10458&quot;, &quot;ece:Z5100&quot;))
</code></pre>
<pre><code>##       hsa:10458       hsa:10458       hsa:10458       hsa:10458       ece:Z5100 
## &quot;path:hsa04520&quot; &quot;path:hsa04810&quot; &quot;path:hsa05130&quot; &quot;path:hsa05135&quot; &quot;path:ece05130&quot;
</code></pre>
</div>
<div class="include-after">
</div>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/components/prism-core.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/plugins/autoloader/prism-autoloader.min.js" defer></script>
</body>
</html>
